### PR TITLE
[LLVM] Add LLVM_full, LLVM_full_assert 19 recipes

### PR DESCRIPTION
**Pull request by Zentrik**

### CODE REVIEW

File: L/LLVM/common.jl [hunk 1]
```diff
@@ -20,6 +20,7 @@ const llvm_tags = Dict(
     v"16.0.6" => "499f87882a4ba1837ec12a280478cf4cb0d2753d", # julia-16.0.6-2
     v"17.0.6" => "0007e48608221f440dce2ea0d3e4f561fc10d3c6", # julia-17.0.6-5
     v"18.1.7" => "ed30d043a240d06bb6e010a41086e75713156f4f", # julia-18.1.7-2
+    v"19.0.0" => "6c8f4c12da7b3ac3d45e9fc9f68e2da7f559cf76", # julia-release/19.x
```
Comment (giordano) [inline L/LLVM/common.jl:23]: Reminder: we need to create the git tag on the commit we will eventually use here.

**Comment by Zentrik** [comment]
Not sure why the assert build can't find bundled

**Comment by giordano** [comment]
You need to either follow symlinks https://docs.binarybuilder.org/stable/reference/#BinaryBuilderBase.DirectorySource or not use symlinks

**Comment by Zentrik** [comment]
> You need to either follow symlinks [docs.binarybuilder.org/stable/reference#BinaryBuilderBase.DirectorySource](https://docs.binarybuilder.org/stable/reference/#BinaryBuilderBase.DirectorySource) or not use symlinks

I don't understand what changed between the llvm 18 build that requires changing anything.

**Comment by giordano** [comment]
https://github.com/Zentrik/Yggdrasil/blob/4a3cf630b50ce9c2d4c06770ebc1f7b3678fbe9e/L/LLVM/LLVM_full_assert%4019/bundled isn't a symlink, but a file whose content is the path of the symlink.

**Comment by Zentrik** [comment]
Looks like it's building nicely with some commits reverted, the reverts are:

- https://github.com/Zentrik/llvm-project/commit/482578d54186312fb5bd210f5c2c5eb53c936749 - Fixes following issue on window
```
ninja: job failed: /opt/bin/x86_64-w64-mingw32-libgfortran5-cxx11/x86_64-w64-mingw32-clang++ --target=x86_64-w64-mingw32 --sysroot=/opt/x86_64-w64-mingw32/x86_64-w64-mingw32/sys-root/ -DGTEST_HAS_RTTI=0 -D_FILE_OFFSET_BITS=64 -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -I/workspace/build/tools/mlir/lib/ExecutionEngine -I/workspace/srcdir/llvm-project/mlir/lib/ExecutionEngine -I/workspace/build/include -I/workspace/srcdir/llvm-project/llvm/include -I/workspace/srcdir/llvm-project/mlir/include -I/workspace/build/tools/mlir/include -remap -D__USING_SJLJ_EXCEPTIONS__ -D__CRT__NO_INLINE -pthread -DMLIR_CAPI_ENABLE_WINDOWS_DLL_DECLSPEC -fvisibility-inlines-hidden -Werror=date-time -Werror=unguarded-availability-new -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -pedantic -Wno-long-long -Wimplicit-fallthrough -Wcovered-switch-default -Wno-noexcept-type -Wnon-virtual-dtor -Wdelete-non-virtual-dtor -Wsuggest-override -Wstring-conversion -ffunction-sections -fdata-sections -Wundef -Werror=mismatched-tags -O3 -DNDEBUG -fvisibility=hidden -fvisibility-inlines-hidden  -fno-exceptions -funwind-tables -fno-rtti -std=c++17 -MD -MT tools/mlir/lib/ExecutionEngine/CMakeFiles/obj.mlir_arm_sme_abi_stubs.dir/ArmSMEStubs.cpp.obj -MF tools/mlir/lib/ExecutionEngine/CMakeFiles/obj.mlir_arm_sme_abi_stubs.dir/ArmSMEStubs.cpp.obj.d -o tools/mlir/lib/ExecutionEngine/CMakeFiles/obj.mlir_arm_sme_abi_stubs.dir/ArmSMEStubs.cpp.obj -c /workspace/srcdir/llvm-project/mlir/lib/ExecutionEngine/ArmSMEStubs.cpp
clang++: warning: argument unused during compilation: '-remap' [-Wunused-command-line-argument]
/workspace/srcdir/llvm-project/mlir/lib/ExecutionEngine/ArmSMEStubs.cpp:37:35: error: dllimport cannot be applied to non-inline function definition
   37 | bool MLIR_ARMSMEABISTUBS_EXPORTED __aarch64_sme_accessible() {
      |                                   ^
/workspace/srcdir/llvm-project/mlir/lib/ExecutionEngine/ArmSMEStubs.cpp:47:40: error: dllimport cannot be applied to non-inline function definition
   47 | sme_state MLIR_ARMSMEABISTUBS_EXPORTED __arm_sme_state() {
      |                                        ^
/workspace/srcdir/llvm-project/mlir/lib/ExecutionEngine/ArmSMEStubs.cpp:52:35: error: dllimport cannot be applied to non-inline function definition
   52 | void MLIR_ARMSMEABISTUBS_EXPORTED __arm_tpidr2_restore() {
      |                                   ^
/workspace/srcdir/llvm-project/mlir/lib/ExecutionEngine/ArmSMEStubs.cpp:56:35: error: dllimport cannot be applied to non-inline function definition
   56 | void MLIR_ARMSMEABISTUBS_EXPORTED __arm_tpidr2_save() {
      |                                   ^
/workspace/srcdir/llvm-project/mlir/lib/ExecutionEngine/ArmSMEStubs.cpp:60:35: error: dllimport cannot be applied to non-inline function definition
   60 | void MLIR_ARMSMEABISTUBS_EXPORTED __arm_za_disable() {
      |                                   ^
5 errors generated.
```
Also, https://github.com/Zentrik/llvm-project/commit/5d9a2e29130f346669233a4fc68e3b05da171a85 to avoid conflict.
- https://github.com/Zentrik/llvm-project/commit/f84efd87ce511008be76ac7fc98bdc87ac32fac6, fixes https://github.com/llvm/llvm-project/issues/106859
- https://github.com/Zentrik/llvm-project/commit/3074f166339e417d7ea24a54c24aa0f6e15a4b59, fixes some error on windows. Also, https://github.com/Zentrik/llvm-project/commit/78d7ef127f40e99a1d9f5a8403d86c4ef9e179d8 to avoid conflict.
- https://github.com/Zentrik/llvm-project/commit/e5a76446eb3881ddc195b8e6c80e28f917c1bb01 fixed issue on windows where building just errored with no message. Also, https://github.com/Zentrik/llvm-project/commit/08d65ddbe40e3553ea997d83adc583707d9ce0fb to avoid conflict.

Also, the problems on non windows seemed to get fixed by these reverts. I would guess the darwin issues were fixed by https://github.com/Zentrik/llvm-project/commit/e5a76446eb3881ddc195b8e6c80e28f917c1bb01.

I guess issues need to be filed and smaller fixes can probably be found.

**Comment by giordano** [comment]
Yes, yesterday I rebased https://github.com/JuliaLang/llvm-project/commits/julia-release/19.x/ on top of 19.1.0.

@Zentrik I see you opened a few tickets upstream about all the issues you faced.  What do you suggest to do?  Wait a few days if there's any movement there?  It has been a few weeks already though.... Should we revert the offending commits in our fork for the time being?  A proper resolution upstream would definitely be a better outcome though, diverging too much from upstream is problematic.

**Comment by Zentrik** [comment]
I'll make some prs upstream and to our fork. Just haven't had the time.

File: L/LLVM/common.jl [hunk 2]
```diff
@@ -20,6 +20,7 @@ const llvm_tags = Dict(
     v"16.0.6" => "499f87882a4ba1837ec12a280478cf4cb0d2753d", # julia-16.0.6-2
     v"17.0.6" => "0007e48608221f440dce2ea0d3e4f561fc10d3c6", # julia-17.0.6-5
     v"18.1.7" => "ed30d043a240d06bb6e010a41086e75713156f4f", # julia-18.1.7-2
+    v"19.1.0" => "376f45c53589090676bdbe7d655813c6d144caaf", # julia-19.1.0-0
```
Comment (giordano) [inline L/LLVM/common.jl:23]:
```suggestion
    v"19.1.0" => "0a1a00654ddded77ba60a361957d8499c0274f30", # julia-19.1.0-0
```
but cloning on Yggdrasil is again taking forever....

**Comment by giordano** [comment]
Build for LLVM_full_assert on `i686-linux-gnu-cxx03` https://buildkite.com/julialang/yggdrasil/builds/13546#01922635-1e02-4a32-bdef-f3341a6d31a6/6-2596
```
[23:22:29] /workspace/srcdir/llvm-project/llvm/include/llvm/CodeGenTypes/MachineValueType.h:43:10: fatal error: llvm/CodeGen/GenVT.inc: No such file or directory
[23:22:29]    43 | #include "llvm/CodeGen/GenVT.inc"
[23:22:29]       |          ^~~~~~~~~~~~~~~~~~~~~~~~
```

**Comment by Zentrik** [comment]
> Build for LLVM_full_assert on `i686-linux-gnu-cxx03` https://buildkite.com/julialang/yggdrasil/builds/13546#01922635-1e02-4a32-bdef-f3341a6d31a6/6-2596
> 
> ```
> [23:22:29] /workspace/srcdir/llvm-project/llvm/include/llvm/CodeGenTypes/MachineValueType.h:43:10: fatal error: llvm/CodeGen/GenVT.inc: No such file or directory
> [23:22:29]    43 | #include "llvm/CodeGen/GenVT.inc"
> [23:22:29]       |          ^~~~~~~~~~~~~~~~~~~~~~~~
> ```

A restart will probably fix it, https://github.com/llvm/llvm-project/issues/107687.

**Comment by giordano** [comment]
We still need to create the tag on the fork repo, but it's all green here! @vchuravy @gbaraldi would you be able to give a quick look?

**Review by maleadt — APPROVED** [review]

**Comment by eschnett** [comment]
LLVM 19.1.1 has been released https://discourse.llvm.org/t/llvm-19-1-1-released/82321

File: L/LLVM/common.jl [hunk 3]
```diff
@@ -20,6 +20,7 @@ const llvm_tags = Dict(
     v"16.0.6" => "499f87882a4ba1837ec12a280478cf4cb0d2753d", # julia-16.0.6-2
     v"17.0.6" => "0007e48608221f440dce2ea0d3e4f561fc10d3c6", # julia-17.0.6-5
     v"18.1.7" => "ed30d043a240d06bb6e010a41086e75713156f4f", # julia-18.1.7-2
+    v"19.1.0" => "0a1a00654ddded77ba60a361957d8499c0274f30", # julia-19.1.0-0
```
Comment (giordano) [inline L/LLVM/common.jl:23]:
```suggestion
    v"19.1.0" => "edb7a552a4a19663899acb6eb47bc795e92cc3e9", # julia-19.1.0-0
```

Comment (Zentrik) [reply L/LLVM/common.jl]: Should this not be 19.1.1

Comment (Zentrik) [reply L/LLVM/common.jl]: Nevermind

Comment (giordano) [reply L/LLVM/common.jl]: I did get it wrong the first time :slightly_smiling_face:

**Comment by Zentrik** [comment]
Did this fail to build or is llvm taking too long to clone?

**Comment by giordano** [comment]
I cancelled them because libgit2 is insufferably slow, I fetched the repos manually with CLI git on the build machine with
```
for agent in {0..15}; do cd /julia/agent-cache/yggy-amdci7.${agent}/yggdrasil/downloads/clones/llvm-project.git-5a9787eb535c2edc5dea030cc221c1d60f38c9f42344f410e425ea2139e233aa&&git fetch; done
```
to make this go (I put the command here so that I remember next time)

**Comment by Zentrik** [comment]
I suspect https://github.com/llvm/llvm-project/issues/110382 is an issue in LLVM 19 as well so probably good to wait a couple days to see if it can get fixed.

File: L/LLVM/LLVM_full@19/build_tarballs.jl [hunk 4]
```diff
@@ -0,0 +1,7 @@
+version = v"19.1.1"
+
+include("../common.jl")
+
+build_tarballs(ARGS, configure_build(ARGS, version; experimental_platforms=true)...;
+               preferred_gcc_version=v"10", preferred_llvm_version=v"18", julia_compat="1.6")
+# Build trigger: 2
```
Comment (giordano) [inline L/LLVM/LLVM_full@19/build_tarballs.jl:7]:
```suggestion
# Build trigger: 3
```

File: L/LLVM/LLVM_full_assert@19/build_tarballs.jl [hunk 5]
```diff
@@ -0,0 +1,7 @@
+version = v"19.1.1"
+
+include("../common.jl")
+
+build_tarballs(ARGS, configure_build(ARGS, version; assert=true, experimental_platforms=true)...;
+               preferred_gcc_version=v"10", preferred_llvm_version=v"18", julia_compat="1.6")
+# Build trigger: 3
```
Comment (giordano) [inline L/LLVM/LLVM_full_assert@19/build_tarballs.jl:7]:
```suggestion
# Build trigger: 4
```

**Comment by giordano** [comment]
@Zentrik the ittapi patch looks broken

**Comment by Zentrik** [comment]
yh https://github.com/JuliaLang/llvm-project/pull/36 breaks the git cloning as ${PROJECT_BINARY_DIR} exists probably

File: L/LLVM/common.jl [hunk 6]
```diff
@@ -20,6 +20,7 @@ const llvm_tags = Dict(
     v"16.0.6" => "499f87882a4ba1837ec12a280478cf4cb0d2753d", # julia-16.0.6-2
     v"17.0.6" => "0007e48608221f440dce2ea0d3e4f561fc10d3c6", # julia-17.0.6-5
     v"18.1.7" => "ed30d043a240d06bb6e010a41086e75713156f4f", # julia-18.1.7-2
+    v"19.1.1" => "3a195326fffa6594ddec8fe9477e848061f42243", # julia-19.1.1-0
```
Comment (giordano) [inline L/LLVM/common.jl:23]:
```suggestion
    v"19.1.1" => "dae03206c1f4dcf812ff27e2860358ccb933856c", # julia-19.1.1-0
```

File: L/LLVM/LLVM_full@19/build_tarballs.jl [hunk 7]
```diff
@@ -0,0 +1,7 @@
+version = v"19.1.1"
+
+include("../common.jl")
+
+build_tarballs(ARGS, configure_build(ARGS, version; experimental_platforms=true)...;
+               preferred_gcc_version=v"10", preferred_llvm_version=v"18", julia_compat="1.6")
+# Build trigger: 3
```
Comment (Zentrik) [inline L/LLVM/LLVM_full@19/build_tarballs.jl:7]:
```suggestion
# Build trigger: 4
```

**Comment by giordano** [comment]
@Zentrik I still need to create the tag, but apart from that, is this now ready to go or are there other things we should wait for?

**Comment by Zentrik** [comment]
> @Zentrik I still need to create the tag, but apart from that, is this now ready to go or are there other things we should wait for?

I think this is ready to go.

**Comment by giordano** [comment]
Thank you so much for all your work!

File: L/LLVM/common.jl (hunk 1 above)
Comment (giordano) [reply L/LLVM/common.jl]: For the record, git tag created: https://github.com/JuliaLang/llvm-project/releases/tag/julia-19.1.1-0

**Pull request merged**